linkerd/service-profiles/src/split.rs

```diff
@@ -0,0 +1,195 @@
+use crate::{Profile, Receiver, Target};
+use futures::{prelude::*, ready};
+use indexmap::IndexSet;
+use linkerd2_addr::Addr;
+use linkerd2_error::Error;
+use linkerd2_stack::{layer, NewService};
+use rand::distributions::{Distribution, WeightedIndex};
+use rand::{rngs::SmallRng, SeedableRng};
+use std::{
+    marker::PhantomData,
+    pin::Pin,
+    task::{Context, Poll},
+};
+use tower::ready_cache::ReadyCache;
+use tracing::debug;
+
+pub fn layer<N, S, Req>() -> impl layer::Layer<N, Service = NewSplit<N, S, Req>> + Clone {
+    let rng = SmallRng::from_entropy();
+    layer::mk(move |inner| NewSplit {
+        inner,
+        rng: rng.clone(),
+        _service: PhantomData,
+    })
+}
+
+#[derive(Debug)]
+pub struct NewSplit<N, S, Req> {
+    inner: N,
+    rng: SmallRng,
+    _service: PhantomData<fn(Req) -> S>,
+}
+
+#[derive(Debug)]
+pub struct Split<T, N, S, Req> {
+    target: T,
+    rx: Receiver,
+    new_service: N,
+    rng: SmallRng,
+    inner: Option<Inner>,
+    services: ReadyCache<Addr, S, Req>,
+}
+
+#[derive(Debug)]
+struct Inner {
+    distribution: WeightedIndex<u32>,
+    addrs: IndexSet<Addr>,
+}
+
+impl<N: Clone, S, Req> Clone for NewSplit<N, S, Req> {
+    fn clone(&self) -> Self {
+        Self {
+            inner: self.inner.clone(),
+            rng: self.rng.clone(),
+            _service: self._service,
+        }
+    }
+}
+
+impl<T, N: Clone, S, Req> NewService<(Receiver, T)> for NewSplit<N, S, Req>
+where
+    S: tower::Service<Req>,
+{
+    type Service = Split<T, N, S, Req>;
+
+    fn new_service(&self, (rx, target): (Receiver, T)) -> Self::Service {
+        Split {
+            rx,
+            target,
+            new_service: self.inner.clone(),
+            rng: self.rng.clone(),
+            inner: None,
+            services: ReadyCache::default(),
+        }
+    }
+}
+
+impl<T, N, S, Req> tower::Service<Req> for Split<T, N, S, Req>
+where
+    Req: Send + 'static,
+    T: AsRef<Addr> + Clone,
+    N: NewService<(Addr, T), Service = S> + Clone,
+    S: tower::Service<Req> + Send + 'static,
+    S::Response: Send + 'static,
+    S::Error: Into<Error>,
+    S::Future: Send,
+{
+    type Response = S::Response;
+    type Error = Error;
+    type Future = Pin<Box<dyn Future<Output = Result<S::Response, Error>> + Send + 'static>>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        let mut update = None;
+        while let Poll::Ready(Some(up)) = self.rx.poll_recv_ref(cx) {
+            update = Some(up.clone());
+        }
```

Comment:
as above, not sure if we need to do this...

Reply:
If the ready cache isn't ready, then we want to be notifed when the profile changes, so we have to drive it to not ready.